src/MySqlConnector/Serialization/MySqlSession.cs

```diff
@@ -35,16 +36,97 @@ public MySqlSession(ConnectionPool pool, int poolGeneration)
 		public ConnectionPool Pool { get; }
 		public int PoolGeneration { get; }
 		public string DatabaseOverride { get; set; }
+		public IPAddress IPAddress => (m_tcpClient?.Client.RemoteEndPoint as IPEndPoint)?.Address;
 
 		public void ReturnToPool() => Pool?.Return(this);
 
-		public bool IsConnected => m_state == State.Connected;
+		public bool IsConnected
+		{
+			get
+			{
+				lock (m_lock)
+					return m_state == State.Connected;
+			}
+		}
+
+		public bool TryStartCancel(MySqlCommand command)
+		{
+			lock (m_lock)
+			{
+				if (m_activeCommand != command)
+					return false;
+				VerifyState(State.Querying, State.CancelingQuery);
+				if (m_state != State.Querying)
+					return false;
+				m_state = State.CancelingQuery;
+			}
+
+			return true;
+		}
+
+		public void DoCancel(MySqlCommand commandToCancel, MySqlCommand killCommand)
+		{
+			lock (m_lock)
+			{
+				if (m_activeCommand != commandToCancel)
+					return;
+
+				try
+				{
+					killCommand.ExecuteNonQuery();
```

Comment:
Set to three seconds [here](https://github.com/mysql-net/MySqlConnector/pull/229/files#diff-7d25f7d140291c4c3a2bfb66fcc7a5a5R225). I don't have any good way of ensuring that failure to cancel is silently ignored (ADO.NET provides no mechanism to report if cancellation worked or not) or verify that it can be retried until `Cancel` succeeds. However the code *looks* good to me.